Error while typing nested pattern match

```scala
class Foo {
  def foo(x: Any): Unit = {
    x match {
      case y: Bar[_] =>
        y.value match {
          case value: Bar[_] => // here x is an instance of Bar[Bar[_]]
          case _ =>
        }
    }
  }
}

class Bar[T] {
  def value: T = ???
}
```

fails while Ychecking with

```
checking Foo.scala after phase MegaPhase{firstTransform, checkReentrant, elimPackagePrefixes}
exception while typing _ of class class dotty.tools.dotc.ast.Trees$Ident # 31
exception while typing y.value match 
  {
    case value @ _: Bar[_] @_ => 
      ()
    case _ => 
      ()
  } of class class dotty.tools.dotc.ast.Trees$Match # 212
exception while typing {
  y.value match 
    {
      case value @ _: Bar[_] @_ => 
        ()
      case _ => 
        ()
    }
} of class class dotty.tools.dotc.ast.Trees$Block # 213
...
...
Exception in thread "main" java.lang.AssertionError: assertion failed: undefined symbol type _
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.transform.TreeChecker$Checker.assertDefined(TreeChecker.scala:190)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedIdent(TreeChecker.scala:307)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1752)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1822)
	at dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:113)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedUnadapted(TreeChecker.scala:269)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1854)
	at dotty.tools.dotc.reporting.trace$.op1$3(trace.scala:32)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:33)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1850)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1862)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typed(TreeChecker.scala:257)
	at dotty.tools.dotc.typer.Typer.typedPattern(Typer.scala:1926)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedCase$1(Typer.scala:1043)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedCase(Typer.scala:1002)
	at dotty.tools.dotc.transform.TreeChecker$Checker.super$typedCase(TreeChecker.scala:409)
	at dotty.tools.dotc.transform.TreeChecker$Checker.$anonfun$typedCase$2(TreeChecker.scala:409)
	at dotty.tools.dotc.transform.TreeChecker$Checker.withDefinedSyms(TreeChecker.scala:180)
	at dotty.tools.dotc.transform.TreeChecker$Checker.typedCase(TreeChecker.scala:408)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedCases$3(Typer.scala:998)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:62)
	at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:78)
```

--- a/compiler/src/dotty/tools/dotc/transform/TreeChecker.scala
+++ b/compiler/src/dotty/tools/dotc/transform/TreeChecker.scala
@@ -177,12 +177,19 @@ class TreeChecker extends Phase with SymTransformer {
       res
     }
 
+    def withPatSyms[T](syms: List[Symbol])(op: => T)(implicit ctx: Context) = {
+      nowDefinedSyms ++= syms
+      val res = op
+      nowDefinedSyms --= syms
+      res
+    }
+
     def assertDefined(tree: untpd.Tree)(implicit ctx: Context) =
       if (
         tree.symbol.maybeOwner.isTerm &&
         !(tree.symbol.is(Label) && !tree.symbol.owner.isClass && ctx.phase.labelsReordered) // labeldefs breaks scoping
       )
-        assert(nowDefinedSyms contains tree.symbol, i"undefined symbol ${tree.symbol}")
+        assert(nowDefinedSyms contains tree.symbol, i"undefined symbol ${tree.symbol} at line " + tree.pos.line)
 
     /** assert Java classes are not used as objects */
     def assertIdentNotJavaClass(tree: Tree)(implicit ctx: Context): Unit = tree match {
@@ -400,7 +407,7 @@ class TreeChecker extends Phase with SymTransformer {
       }
 
     override def typedCase(tree: untpd.CaseDef, pt: Type, selType: Type, gadtSyms: Set[Symbol])(implicit ctx: Context): CaseDef = {
-      withDefinedSyms(tree.pat.asInstanceOf[tpd.Tree].filterSubTrees(_.isInstanceOf[ast.Trees.Bind[_]])) {
+      withPatSyms(tpd.patVars(tree.pat.asInstanceOf[tpd.Tree])) {
         super.typedCase(tree, pt, selType, gadtSyms)
       }
     }

--- a/tests/pos/i4198.scala
+++ b/tests/pos/i4198.scala
@@ -1,0 +1,15 @@
+class Foo {
+  def foo(x: Any): Unit = {
+    x match {
+      case y: Bar[_] =>
+        y.value match {
+          case value: Bar[_] => // here x is an instance of Bar[Bar[_]]
+          case _ =>
+        }
+    }
+  }
+}
+
+class Bar[T] {
+  def value: T = ???
+}

